src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILEInputStreamExtensions.kt

```diff
@@ -0,0 +1,90 @@
+package io.ionic.libs.ionfilesystemlib.helper.common
+
+import android.util.Base64
+import io.ionic.libs.ionfilesystemlib.model.IONFILEEncoding
+import io.ionic.libs.ionfilesystemlib.model.IONFILEReadInChunksOptions
+import io.ionic.libs.ionfilesystemlib.model.IONFILEReadOptions
+import java.io.InputStream
+import java.io.InputStreamReader
+
+/**
+ * Reads the entire contents of an [InputStream] with options
+ *
+ * @param options the options for configuring how to read from the stream
+ * @return the full contents of the stream
+ */
+internal fun InputStream.readFull(options: IONFILEReadOptions): String =
+    if (options.encoding is IONFILEEncoding.WithCharset) {
+        val reader =
+            InputStreamReader(this, options.encoding.charset)
+        reader.use { reader.readText() }
+    } else {
+        val byteArray = this.readBytes()
+        Base64.encodeToString(byteArray, Base64.NO_WRAP)
+    }
+
+/**
+ * Reads the contents of an [InputStream] in chunks.
+ *
+ * This method is suspend, and should be called from a non-main thread (e.g. using Dispatchers.IO)
+ *
+ * @param options for reading from the stream, including the chunk size to return
+ * @param bufferSize the size of the buffer for reading from the stream.
+ *  This is different from the chunk size, and should be a value that aligns with the OS page size
+ *  The buffer size may alter the chunkSize value to be used; refer to [IONFILEReadInChunksOptions]
+ * @param onChunkRead suspend function in which to return each chunk that was read
+ */
+internal suspend fun InputStream.readByChunks(
+    options: IONFILEReadInChunksOptions,
+    bufferSize: Int,
+    onChunkRead: suspend (String) -> Unit,
+) {
+    val chunkSize = minOf(options.chunkSize, available())
+        .coerceAtLeast(bufferSize)
+        .let {
+            if (options.encoding == IONFILEEncoding.Base64) {
+                // make chunk size the nearest highest multiple of 3, to not add padding to the end
+                //  this is so that multiple chunks can be concatenated and decoded correctly
+                it - (it % 3) + 3
+            } else {
+                it
+            }
+        }
+    var bytesRead: Int
+    var readCount = 0
+    do {
```

Comment:
This code could all be put inside of a `use` block from `InputStream` to make sure it is closed properly at the end, right?

Reply:
Yeah that makes total sense.